Call getAccountBalance for all wallets at once on tip change

Anviking opened the pull request:
# Issue Number

#2005 

# Overview

- [x] Only send the LSQ query `GetFilteredDelegationsAndRewardAccounts` when the tip changes
- [x] Query _all_ wallet reward accounts in _the same_ `GetFilteredDelegationsAndRewardAccounts`

Listing stake pools with 10 wallets now takes ~5 seconds instead of 2-3 minutes!

- [x] Make the implementation easier to follow and actually sane

# Comments

- [ ] Not completely sure this doesn't mess up some some intricate interactions between the reward balance state and utxo state. The integration tests pass. But maybe there are some rollback scenarios which could lead to the reward balance being incorrectly reported for a while… Would be nice to think through properly and synchronise them. 

I think this happens on tip changes:

1. Wallet worker updates UTxO
2. Tip worker re-feches balances and writes to TVar
3. (I think on next tip change) Wallet reads the balance from the TVar and writes it to Sqlite
4. New reward balance is queryable from API

With this PR, I wonder if it may not risk being out of date for one more tip-change
<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Batch getAccountBalance from multiple workers together

getAccountBalance is called by each wallet worker, I think on each
restoration step (which is often).

For some reason, sending multiple getAccountBalance queries at the same
time slows it down from e.g. 0.002s to 40s.

The LSQ query even supports querying multiple accounts in a single
query, so let's use it!

This commit makes getAccountBalance merely lookup in a Map in a TVar,
which is updated with a single query on every tip-change.

Somewhat crude implementation so far.

I'm yet not familiar with the code that is calling getAccoutBalance. I
suspect there are layers of indirection that can be removed.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -328,6 +349,8 @@ withNetworkLayer tr np addrInfo versionData action = do
     connectNodeTipClient
         :: HasCallStack
         => RetryHandlers
+        -> TVar IO (Map W.ChimericAccount (Maybe W.Coin))
```
What about using `0` as a default value instead of `Nothing`. We do already default to `0` when it comes to the reward balance and it is quite plausible / will not cause any particular harm to report a balance as 0 until we fetch it.

Anviking replied:
I wanted to keep "we already default to 0" separate from the core caching logic.

With the slight refactoring, I made new entries get added to a separate `toBeObserved :: Set`, instead of `Nothing` values inside the `Map`, but the "query" still returns `Maybe W.Coin` like here, requiring an explicit `fromMaybe`.

KtorZ reviewed (commented):
I like the principle here of fetching them all in one query. We need to make sure however to clean accounts from the list when wallets are deleted. Otherwise, we end up fetching unused account over and over. We need some form "pub / sub" mechanism.

Anviking committed:
Add Observer abstraction for fetching rewards

Seems slightly nicer.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -984,15 +1092,14 @@ instance TPraosCrypto sc => ToText (NetworkLayerLog sc) where
+        MsgGetRewardAccountBalance tip accts -> T.unwords
             [ "Querying the reward account balance for"
-            , pretty acct
+            , fmt $ listF accts
```
With 10 wallets:
```
[cardano-wallet.network:Info:15] [2020-08-17 10:37:49.31 UTC] Querying the reward account balance for  [1ed0bdd4, 2a98402c, 5237aa81, 634bea28, 69e2c672, 75765c1e, 84132baa, c662444a, d55deafa, f4ea9095] at 70617265<-[c90c882f-6094377#4570261]
```

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -767,6 +761,92 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
+            , stopObserving = \k ->
+                atomically $ do
+                    modifyTVar' toBeObservedVar (Set.delete k)
+                    modifyTVar' cacheVar (Map.delete k)
```
Integration tests / [ STAKE_POOLS_JOIN_01](https://github.com/input-output-hk/cardano-wallet/blob/f66a86e01b89a0fbe32df6623f3923fc3d9fd6bd/lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs#L152-L307) pass, meaning this should work.

But if this `Observer` abstraction is something we want to keep, it would be nice to add more extensive unit tests. `newRewardBalanceFetcher` would probably have to be made more abstract. (Because the record itself provides very little structure and can't be tested)

I also guess there is similarities between this and the `WorkerRegistry`, that perhaps could be unified in the future.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -771,6 +761,92 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
+        -> TVar IO (Set W.ChimericAccount)
+        -> Tip (CardanoBlock sc)
+        -> IO ()
+    refresh cacheVar toBeObservedVar tip = do
```
I find it surprising that `refresh` is not part of the `Observer` "interface". The function clearly has a privileged access to both TVar and make strong assumption about how they are manipulated. So to me, it seems like it belong to the same abstraction and having it as a separate function is a bit confusing.

Anviking replied:
To me, `Observer` is concerned about _subscribing to_ and _accessing_ values. It theoretically be passed around and escape the `Network` module. 

`refresh` is coupled to the _creation_ of the `Observer`. It's more like an implementation detail, specific to the `Network` module.

If say the wallet registry were to have access to an `Observer`, it should not be allowed to manually call `refresh`.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -771,6 +761,92 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
+          , Tip (CardanoBlock sc) -> IO ()
+            -- Call on tip-change to refresh
+          )
+newRewardBalanceFetcher tr gp queryRewardQ = do
```
It'd be nice to be able to test this bit of logic in isolation. This is currently hard because the management of the keys / accounts is mixed with the logic for fetching the reward. I'd suggest to replace the `ChimericAccount` with a `Ord k => k` and to replace `Coin` with an abstract `a`, and pass a function that yield `a` as argument. So, something like:


 ```hs
data Observer m key value = Observer
    { startObserving :: key -> m ()
    , stopObserving :: key -> m ()
    , query :: key -> m (Maybe value)
    , refresh :: tip -> m tip
    }

newObserver 
    :: Ord key
    => (tip -> [key] -> m (Map key value))
    -> Observer m key value
newObserver getValues = ...
```

KtorZ replied:
This way, we can separate the network logic from the observer logic and test them separately (at least, test the observer).

Anviking replied:
> It'd be nice to be able to test this bit of logic in isolation. 

Yes 👍  I alluded to it in https://github.com/input-output-hk/cardano-wallet/pull/2034#discussion_r471493798, but didn't want to explore it before hearing your feedback.

Anviking committed:
Separate newRewardBalanceFetcher and newObserver

Anviking committed:
Add basic unit tests for Observer

Anviking committed:
Track registrations separately from cache

Before this commit keys would start in `toBeObserved`. They would then
get passed to the `fetch` function. If the supplied `fetch` function
chose to not include the same keys in the resulting map of values, they
would silently dissappear.

This way, `newObserver` takes full responsibility of maintaining the set
of keys which should be observed, by replacing the TVars (cache,
toBeObserved) with (cache, observedKeys)

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -771,6 +766,121 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
+            { startObserving = \k -> do
+                atomically $ do
+                    modifyTVar' observedKeysVar (Set.insert k)
+                traceWith tr $ MsgAddedObserver k
```
Since "startObserving" is called every time `getAccountBalance` is called, this would log a line `MsgAddedObserver` in a rather misleading way and much more than it needs. We could use something like `Set.alterF` and return `Just _` or `Nothing` depending on whether the element was added (and log accordingly).

Anviking replied:
Yup. It's connected to `nullTracer` now, so one doesn't notice, but should connect to the real tracer and change this.

Anviking committed:
Fixup: whitespace

KtorZ commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs:
```diff
@@ -47,6 +54,54 @@ spec = describe "NetworkLayer regression test #1708" $ do
+    describe "Observer" $ do
+        describe "(query k) with typical use" $ beforeAll mockObserver $ do
+            -- Using monadic-property tests /just/ for the sake of testing with
+            -- multiple keys seem worthless.
```
It'd be nice to test with multiple keys, since this is the main _point_ of this PR actually. Coming up with some properties would help testing possible edge-cases that aren't caught here, although may be a bit tricky here. At least, we should add two additional test scenarios:

- registering the same key twice
- registering multiple keys

Anviking replied:
I should say, I’m much more worried about this causing further balance discrepancies than that it doesn’t handle registering the same key twice, or multiple keys, etc (though sure, good to test).

Anviking committed:
More extensive testing and logging

KtorZ commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs:
```diff
@@ -47,6 +67,155 @@ spec = describe "NetworkLayer regression test #1708" $ do
+                            [ MsgAddedObserver k
+                            , MsgWillFetch $ Set.singleton k
+                            , MsgDidFetch $ Map.singleton k v
+                            ]
```
I don't get this test, unless this test depends on the previous ones which would be quite bad / misleading. Tests should work in isolation.

Anviking replied:
I wanted to experiment with slightly different usage of `describe`/`it` for more readable test output.

Here we get:
```haskell
        when refresh fails
          (query k) returns the existing v
          only MsgWillFetch is traced
```
which I think is quite neat.

(And in total:)
```
      typical use
        startObserving
          (query k) returns Nothing before startObserving
          (query k) returns v after (startObserving k >> refresh)
          traced MsgAddedObserver, MsgWillFetch, MsgDidFetch
        calling startObserving a second time
          (query k) is still v
        when refresh fails
          (query k) returns the existing v
          only MsgWillFetch is traced
        stopObserving
          makes (query k) return Nothing
```

With no shared state, the whole `typical use` tree would be a single `it`.

I'll admit it gets a bit weird here though, where the `refresh False` belongs to the whole `when refresh fails` tree, but is run in the first `it`. (We cannot have it as a `runIO` or `beforeAll`, because we cannot access the `refresh` function from outside an `it`).

We'd need some
```haskell
mapBeforeAll :: (a -> IO b) -> SpecWith b -> Spec a
```
which doesn't exist

Anviking replied:
We _could_ probably add a `Arbitrary (Observer, refresh)`, and _do_ run most of these checks in isolation though 🤔

KtorZ replied:
I admit that the separation is neat, but it is puzzling. So at the very least, there should be a clear note as comment explaining basically what you just explained here.

Anviking committed:
Add comments explaining statefulness and small it blocks

Anviking committed:
Minor documentation comment adjustments

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -770,6 +764,151 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
+        Observer
+            { startObserving = \k -> do
+                wasAdded <- atomically $ do
+                    notAlreadyThere <- Set.notMember k <$> readTVar observedKeysVar
```
Seems `Set.alterF` doesn't exist with our version of `containers` , and I didn't figure out how it worked, but think a separate `readTVar` inside `atomically` is perfectly fine.
